### PR TITLE
feat: support querying properties of Blocks and Events

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Refactor how dataframes are pulled from the query plugin system (pulls attributes from objects instead of using Pydantic's `.dict` method)

This allows querying objects that don't appear in a model's Pydantic schema

Also refactored how `validate_and_expand_columns` works so that you pass a model instead of a list of fields

### How I did it
slight abuse of MRO and `vars`

### How to verify it

`df = chain.blocks.query("transactions")`

### Checklist

- [x] All changes are completed
- [x] New test cases have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by fubuloubu** [comment]
With this PR, can do stuff like this now:

```py
In [1]: df = chain.blocks.query("transactions", start_block=-10)
INFO: Cache database has not been initialized

# Sum of ether transfers for all transactions in a block
In [2]: df.transactions.map(lambda tl: sum(t.value for t in tl) / 1e18)
Out[2]: 
0     433.829804
1     341.989568
2      15.939023
3    1613.561825
4       8.650647
5       0.830293
6      75.449397
7      16.537289
8       1.270925
9       1.728133
Name: transactions, dtype: float64
```

File: src/ape/api/query.py [hunk 1]
```diff
@@ -15,21 +15,41 @@
...
     if len(columns) == 1 and columns[0] == "*":
-        return all_columns
+        # NOTE: By default, only pull explicit fields
+        #       (because they are cheap to pull, but properties might not be)
+        return list(Model.__fields__)
 
     else:
-        if len(set(columns)) != len(columns):
+        deduped_columns = set(columns)
+        if len(deduped_columns) != len(columns):
```
Comment (johnson2427) [inline src/ape/api/query.py:26]: considering deduped_columns would inherently remove duplicates, should we just fire off a warning letting the user know that they put duplicate fields in and we removed them? No need to raise here if we do so

File: src/ape/managers/chain.py [hunk 2]
```diff
@@ -139,17 +140,15 @@ def query(
             )
 
         query = BlockQuery(
-            columns=list(self.head.__fields__),
+            columns=columns,
```
Comment (johnson2427) [inline src/ape/managers/chain.py:143]: Good work on this, love that we were able to clean this up

**Review by johnson2427 — APPROVED** [review]
LGTM at the moment, trying to get my environment to pass these tests, but this is definitely something on my end

File: src/ape/api/query.py [hunk 3]
```diff
@@ -15,21 +15,41 @@
...
             raise ValueError(f"Duplicate fields in {columns}")
```
Comment (fubuloubu) [inline src/ape/api/query.py:27]:
```suggestion
```

File: src/ape/api/query.py [hunk 4]
```diff
@@ -15,21 +15,41 @@
...
+                for field in vars(cls)
+                if not field.startswith("_")
+                and isinstance(vars(cls)[field], (property, cached_property))
+            }
+        )
+
+        for c in deduped_columns - all_columns:
+            raise ValueError(f"Unrecognized field '{c}', must be one of {all_columns}")
 
     return columns
```
Comment (fubuloubu) [inline src/ape/api/query.py:46]:
```suggestion
        return list(all_columns.union(deduped_columns))
```

File: src/ape/api/query.py [hunk 5]
```diff
@@ -15,21 +15,41 @@
...
+                for cls in Model.__mro__
+                if issubclass(cls, BaseInterfaceModel) and cls != BaseInterfaceModel
+                for field in vars(cls)
+                if not field.startswith("_")
+                and isinstance(vars(cls)[field], (property, cached_property))
+            }
+        )
+
+        for c in deduped_columns - all_columns:
+            raise ValueError(f"Unrecognized field '{c}', must be one of {all_columns}")
```
Comment (fubuloubu) [inline src/ape/api/query.py:44]:
```suggestion
        if len(deduped_columns - all_columns) > 0:
            unrecognized = "', '".join(deduped_columns - all_columns)
            logger.warning(f"Unrecognized fields '{unrecognized}', must be one of {all_columns}")
```

**Review by fubuloubu — COMMENTED** [review]
@johnson2427 so basically more like this? (e.g. just warning, not error, and pruning unrecognized fields for safety)

File: src/ape/api/query.py (hunk 5 above)
Comment (johnson2427) [reply src/ape/api/query.py:44]: This is perfect

**Review by unparalleled-js — DISMISSED** [review]
Looks really good!

**Comment by fubuloubu** [comment]
Cool side effect! This works:

```py
>>> yfi = Contract("0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e")
>>> df = yfi.Transfer.query("from", "to", "value", start_block=-1000)
                                          from                                          to                value
0   0x89B8B956A5A6fAac3532E353575FEe2E1D52014F  0xA9D1e08C7793af67e9d92fe308d5697FB81d3E43    19897670000000000
1   0x088ee5007C98a9677165D78dD2109AE4a3D04d0C  0x0000000000007F150Bd6f54c40A34d7C3d5e9f56    92110183393781795
2   0x0000000000007F150Bd6f54c40A34d7C3d5e9f56  0xFEB4acf3df3cDEA7399794D0869ef76A6EfAff52    92110183393815472
3   0x088ee5007C98a9677165D78dD2109AE4a3D04d0C  0x56178a0d5F301bAf6CF3e1Cd53d9863437345Bf9   613925044775865623
4   0xC26b89A667578ec7b3f11b2F98d6Fd15C07C54ba  0xF2F400C138F9fb900576263af0BC7fCde2B1b8a8  1117463633192112032
..                                         ...                                         ...                  ...
91  0x21a31Ee1afC51d94C2eFcCAa2092aD1028285549  0x000000000dFDe7deaF24138722987c9a6991e2D4  9999590000000000000
92  0xA9D1e08C7793af67e9d92fe308d5697FB81d3E43  0x257201ca66b38dC2C92158E9Dd5502B8B9D8dEB1   324779970000000000
93  0xE68c1d72340aEeFe5Be76eDa63AE2f4bc7514110  0x24902AA0cf0000a08c0EA0b003B0c0bF600000E0    71519175236582310
94  0x24902AA0cf0000a08c0EA0b003B0c0bF600000E0  0xBA12222222228d8Ba445958a75a0704d566BF2C8    71519175236582310
95  0x56178a0d5F301bAf6CF3e1Cd53d9863437345Bf9  0x088ee5007C98a9677165D78dD2109AE4a3D04d0C   593216490757629592

[96 rows x 3 columns]
```

**Review by unparalleled-js — APPROVED** [review]
lgtm!

**Pull request merged**